CDCgov/prime-reportstream
# #2225 - Include qualityFilter information in response

This PR ...
- [x] includes destinations filtered out based on quality in response.
- [x] lists the items that have been filtered for each destination in response.

**Test Steps:**
With SmokeTest:
1. run `./gradlew testSmoke`
2. Look at results for Quality Filter tests, they will have entries for `filteredReportRows`

By Hand:
1. Submit files that should have rows or whole entries filtered out.
2. See that response contains the correct `filteredReportRows`
3. See that the `secondary` does _not_ have `filteredReportRows`


[ReportStream HL7 Filtered Rows.postman_collection.json.zip](https://github.com/CDCgov/prime-reportstream/files/7443761/ReportStream.HL7.Filtered.Rows.postman_collection.json.zip)

**For Review:**
- Clarity
- Separation of concerns
- Unexpected consequences

**Fixes:**
#2225

**Example response:**
```json
{
    "id": "d9cf2c2e-b138-49fb-a0e6-7bd00a4c638c",
    "timestamp": "2021-10-29T17:06:08.857916Z",
    "topic": "covid-19",
    "reportItemCount": 3,
    "destinations": [
        {
            "organization": "California Department of Public Health",
            "organization_id": "ca-dph",
            "service": "elr",
            "filteredReportRows": [
                "For ca-dph.elr, qualityFilter hasValidDataFor, [message_id, equipment_model_name, specimen_type, test_result, patient_last_name, patient_first_name, patient_dob] filtered out Rows 1 reducing the Item count from 3 to 2."
            ],
            "sending_at": "2021-10-29T13:07-04:00",
            "itemCount": 2
        },
        {
            "organization": "California Department of Public Health",
            "organization_id": "ca-dph",
            "service": "elr-secondary",
            "sending_at": "2021-10-29T13:07-04:00",
            "itemCount": 1
        }
    ],
    "destinationCount": 2,
    "warningCount": 3,
    "errorCount": 0,
    "errors": [],
    "warnings": [
        {
            "scope": "ITEM",
            "id": "",
            "details": "Report 1: Timestamp for ORC-15 - order_test_date should provide more precision. Should be formatted as YYYYMMDDHHMM[SS[.S[S[S[S]+/-ZZZZ"
        },
        {
            "scope": "ITEM",
            "id": "",
            "details": "Report 2: Timestamp for ORC-15 - order_test_date should provide more precision. Should be formatted as YYYYMMDDHHMM[SS[.S[S[S[S]+/-ZZZZ"
        },
        {
            "scope": "ITEM",
            "id": "",
            "details": "Report 3: Timestamp for ORC-15 - order_test_date should provide more precision. Should be formatted as YYYYMMDDHHMM[SS[.S[S[S[S]+/-ZZZZ"
        }
    ],
    "consolidatedErrors": [],
    "consolidatedWarnings": [
        {
            "message": "Report 3: Timestamp for ORC-15 - order_test_date should provide more precision. Should be formatted as YYYYMMDDHHMM[SS[.S[S[S[S]+/-ZZZZ",
            "rows": ""
        },
        {
            "message": "Report 2: Timestamp for ORC-15 - order_test_date should provide more precision. Should be formatted as YYYYMMDDHHMM[SS[.S[S[S[S]+/-ZZZZ",
            "rows": ""
        },
        {
            "message": "Report 1: Timestamp for ORC-15 - order_test_date should provide more precision. Should be formatted as YYYYMMDDHHMM[SS[.S[S[S[S]+/-ZZZZ",
            "rows": ""
        }
    ]
}
```

```json
{
    "id": "c07a22fe-30ae-4366-8e88-470873313712",
    "timestamp": "2021-10-29T17:06:35.530668Z",
    "topic": "covid-19",
    "reportItemCount": 1,
    "destinations": [
        {
            "organization": "California Department of Public Health",
            "organization_id": "ca-dph",
            "service": "elr",
            "filteredReportRows": [
                "For ca-dph.elr, qualityFilter hasValidDataFor, [message_id, equipment_model_name, specimen_type, test_result, patient_last_name, patient_first_name, patient_dob] filtered out Rows 0 reducing the Item count from 1 to 0."
            ],
            "sending_at": "never - all items filtered out",
            "itemCount": 0
        },
        {
            "organization": "California Department of Public Health",
            "organization_id": "ca-dph",
            "service": "elr-secondary",
            "sending_at": "2021-10-29T13:07-04:00",
            "itemCount": 1
        }
    ],
    "destinationCount": 2,
    "warningCount": 1,
    "errorCount": 0,
    "errors": [],
    "warnings": [
        {
            "scope": "ITEM",
            "id": "",
            "details": "Report 1: Timestamp for ORC-15 - order_test_date should provide more precision. Should be formatted as YYYYMMDDHHMM[SS[.S[S[S[S]+/-ZZZZ"
        }
    ],
    "consolidatedErrors": [],
    "consolidatedWarnings": [
        {
            "message": "Report 1: Timestamp for ORC-15 - order_test_date should provide more precision. Should be formatted as YYYYMMDDHHMM[SS[.S[S[S[S]+/-ZZZZ",
            "rows": ""
        }
    ]
}
```
